kyo-prelude/shared/src/main/scala/kyo2/kernel/Boundary.scala

```diff
@@ -1,51 +1,29 @@
 package kyo2.kernel
 
 import internal.*
+import kyo.Tag
+import scala.annotation.tailrec
 import scala.quoted.*
 
-final class Boundary[Ctx, S] private (dummy: Unit) extends AnyVal:
-
-    inline def apply[A, B, S2, S3](inline v: => A < (Ctx & S & S2))(
-        inline f: A < (S & S2) => B < S3
-    )(using inline _frame: Frame): B < (Ctx & S3) =
-        new KyoDefer[B, Ctx & S3]:
-            def frame = _frame
-            def apply(dummy: Unit, context: Context)(using safepoint: Safepoint) =
-                val state = safepoint.save(context)
-                f(Effect.defer(boundaryLoop(v, state)))
-
-    inline def apply[A, B, S2, S3](seq: Seq[A < (Ctx & S & S2)])(
-        inline f: Seq[A < (S & S2)] => B < S3
-    )(using inline _frame: Frame): B < (Ctx & S3) =
-        new KyoDefer[B, Ctx & S3]:
-            def frame = _frame
-            def apply(dummy: Unit, context: Context)(using safepoint: Safepoint) =
-                val state = safepoint.save(context)
-                f(seq.map(item => boundaryLoop(item, state)))
-
-    private def boundaryLoop[A, S2](v: A < (Ctx & S & S2), state: Safepoint.State)(using Frame): A < (S & S2) =
-        v match
-            case <(kyo: KyoSuspend[IX, OX, EX, Any, A, S] @unchecked) =>
-                new KyoContinue[IX, OX, EX, Any, A, S & S2](kyo):
-                    def frame = summon[Frame]
-                    def apply(v: OX[Any], context: Context)(using Safepoint) =
-                        val parent = Safepoint.local.get()
-                        Safepoint.local.set(Safepoint(parent.depth, parent.interceptor, state))
-                        val r =
-                            try kyo(v, state.context)
-                            finally Safepoint.local.set(parent)
-                        boundaryLoop(r, state)
-                    end apply
-            case <(kyo) =>
-                kyo.asInstanceOf[A]
+final class Boundary[Ctx, +S] private (dummy: Unit) extends AnyVal:
 
 end Boundary
 
-object Boundary:
+private[kyo2] object Boundary:
+
+    extension [Ctx, S](boundary: Boundary[Ctx, S])
+        private[kyo2] inline def apply[A, S2](inline f: (Trace, Context) => A < S2)(using inline _frame: Frame): A < (S & S2) =
+            new KyoDefer[A, S & S2]:
+                def frame = _frame
+                def apply(v: Unit, context: Context)(using safepoint: Safepoint) =
+                    f(safepoint.saveTrace(), context)
 
     inline given derive[Ctx, S]: Boundary[Ctx, S] = ${ boundaryImpl[Ctx, S] }
 
-    def apply[Ctx, S](using boundary: Boundary[Ctx, S]): Boundary[Ctx, S] = boundary
```

Comment:
This `apply` method was conflicting with the new extension method. Since it was used only for testing, I'm using `Boundary.derive` directly instead in tests.